syz-manager/snapshot.go

```diff
@@ -0,0 +1,172 @@
+// Copyright 2024 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package main
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/google/flatbuffers/go"
+	"github.com/google/syzkaller/pkg/flatrpc"
+	"github.com/google/syzkaller/pkg/fuzzer/queue"
+	"github.com/google/syzkaller/pkg/log"
+	"github.com/google/syzkaller/vm"
+)
+
+func (mgr *Manager) snapshotLoop() {
+	queue.StatNumFuzzing.Add(mgr.vmPool.Count())
+
+	errc := make(chan error)
+	for index := 0; index < mgr.vmPool.Count(); index++ {
+		index := index
+		go func() {
+			errc <- mgr.snapshotVM(index)
+		}()
+	}
+	log.Fatal(<-errc)
+}
+
+func (mgr *Manager) snapshotVM(index int) error {
+	inst, err := mgr.vmPool.Create(index)
+	if err != nil {
+		return err
+	}
+	executor, err := inst.Copy(mgr.cfg.ExecutorBin)
+	if err != nil {
+		return err
+	}
+	// All network connections (including ssh) will break once we start restoring snapshots.
+	// So we start a background process and log to /dev/kmsg.
+	cmd := fmt.Sprintf("nohup %v exec snapshot 1>/dev/null 2>/dev/kmsg </dev/null &", executor)
+	if _, _, err := inst.Run(time.Hour, mgr.reporter, cmd); err != nil {
+		return err
+	}
+
+	builder := flatbuffers.NewBuilder(0)
+	var envFlags flatrpc.ExecEnv
+	for first := true; ; first = false {
+		queue.StatExecs.Add(1)
+		req := mgr.source.Next()
+		if first {
+			envFlags = req.ExecOpts.EnvFlags
+			if err := mgr.snapshotSetup(inst, builder, envFlags); err != nil {
+				return err
```

Comment:
Nit: if it failed due to a kernel crash (this may rarely happen on some unstable kernels), we may want to record that as a crash and not abort the whole fuzzing. But that's not critical for now.

Reply:
Done.
I've restructured code to start failed instances on errors, rather than hard fail.